### PR TITLE
Return predicate failed message in unified resource requests

### PR DESCRIPTION
**Pull request by avatus**
This will properly return a match error during iteration or fetching by IDs so we can parse it on the UI. Currently, we "fail silently".

### CODE REVIEW

**Comment by avatus** [comment]
I tested locally pretty heavily but i'd appreciate an extra round of testing to make sure I didn't forget any edge cases

File: web/packages/teleport/src/UnifiedResources/Resources.tsx [hunk 1]
```diff
@@ -278,9 +278,14 @@ export function Resources() {
           <ErrorBoxInternal>
             <Danger>
               {attempt.statusText}
-              <Box flex="0 0 auto" ml={2}>
-                <ButtonLink onClick={onRetryClicked}>Retry</ButtonLink>
-              </Box>
+              {/* we don't want them to try another request with failed predicate expression, it will just fail again. */}
+              {!attempt.statusText.includes(
+                'failed to parse predicate expression'
```
Comment (zmb3) [inline web/packages/teleport/src/UnifiedResources/Resources.tsx:283]: This will break as soon as the error message changes.
To avoid this, I would put a comment in the Go code where that message is defined, warning devs not to change it.

File: lib/services/unified_resource.go [hunk 2]
```diff
@@ -305,14 +307,18 @@ func (c *UnifiedResourceCache) GetUnifiedResourcesByIDs(ctx context.Context, ids
 				continue
 			}
 			resource := cache.resources[res.Value]
-			if matched := matchFn(resource); matched {
+			match, err := matchFn(resource)
+			if err != nil {
+				return trace.Wrap(err)
```
Comment (zmb3) [inline lib/services/unified_resource.go:312]: Do we have any existing tests that cover this new behavior where a match fn can fail?

File: web/packages/teleport/src/UnifiedResources/Resources.tsx (hunk 1 above)
Comment (avatus) [reply web/packages/teleport/src/UnifiedResources/Resources.tsx:284]: Added a comment and added it to the test. which really feels backwards like I'm super hardening a fickle conditional anyway but, we can refactor the frontend `useAttempt` to return a status code that we can work with instead in a future PR. I added a TODO

Comment (avatus) [reply web/packages/teleport/src/UnifiedResources/Resources.tsx:284]: I agree. it returns a 400 currently. we just need to refactor the `useAttempt` hook and its callsites to include this statusCode on error which might blow up the PR a bit (all of web uses it)

Comment (avatus) [reply web/packages/teleport/src/UnifiedResources/Resources.tsx:284]: Actually, this refactor might not be too big. I think I can squeeze it into this PR. hold off on further review

Comment (avatus) [reply web/packages/teleport/src/UnifiedResources/Resources.tsx]: Thanks for pushing on this. A similar refactor happened with the async code that Connect uses `useAsync` and I remember it being a large thing. For `useAttempt` it turned out to be easy since all the request stuff is updated manually by the caller. I added an optional `statusCode` field and didn't have to update any of the other callsites (although I'm sure I can find some that would benefit from this change. Thats for another day).

**Comment by avatus** [comment]
Ultra friendly ping @rudream @ryanclark

**Review by zmb3 — APPROVED** [review]

**Review by rudream — APPROVED** [review]
